_posts/2016-03-29-Reactivecocoa-+-MVVM.md

```diff
@@ -0,0 +1,96 @@
+
+## MVVM
+![]({{ site.baseurl }}/img/Delarge/MVVMPattern.png)
+
+Model-View-ViewModel, ou MVVM como é conhecido, trata-se de um design pattern, que consiste em separar lógicas de UI, da lógica de negócios ,
+tornando as aplicações mais faceis de serem desenvolvidas e testadas. Reúne todos os dados brutos que podem vir de banco de dados, webservices e etc, aplica lógicas, e formata estes dados para apresentação em uma viewcontroller. Expõe via properties somente a informação que a viewcontroller precisa saber para efetuar o trabalho de mostrar os dados em uma view .
+
+## MVVM e binding de dados
+
+![]({{ site.baseurl }}/img/Delarge/MVVMReactiveCocoa.png)
+
+No contexto deste post, de utilizar ReactiveCocoa somente para binding de dados, usando a arquitetura MVVM, o ReactiveCocoa executa uma função muito específica. Ele fornece uma espécie de "ponte" da viewmodel para a view,  monitorando todas as mudanças realizadas no modelo de dados, e mapeando estas mudanças para as properties da viewmodel efetuando qualquer lógica de negócios necessária .
+Para dar um exemplo, imagine que nosso modelo de dados, contenha uma property NSDate chamada de dateAdded, e nós iremos monitorar suas mudanças, para  atualizar outra property também chamada de dateAdded, mas esta, do nosso viewmodel. A property dateAdded do nosso modelo, como dito anteriormente, será uma NSDate enquanto a do nosso viewmodel será uma NSString. O binding irá parecer com algo descrito abaixo :
+
+~~~objc
+@property (nonatomic) NSString *dateAdded;
+@property (nonatomic) Model *model;
+
+RAC(self,dateAdded)=[RACObserve(self.model,dateAdded)
+	map:^(NSDate *date){
+	 return [[ViewModel dateFormatter] stringFromDate:date];
+ }];
+~~~
+
+dateFormatter é um metodo de classe em nosso viewmodel, que faz uso de cache de uma instância NSDateFormatter que estará pronta para reuso (NSDateFormatter são muito custosas para se criar a todo tempo). Então nossa viewcontroller poderá monitorar todas as mudanças da nossa property dateAdded da nossa viewModel, podendo exibí-la em uma label,  por exemplo.
+
+
+~~~objc
+RAC(self.label,text) = RACObserve(self.viewModel,dateAdded);
+~~~
+
+Aqui nós estamos abstraindo toda a lógica de se transformar um NSDate para uma NSString para a nossa viewmodel, onde nós podemos escrever testes unitários para ela . Pode parecer um pouco confuso neste exemplo e um tanto quanto trivial como podemos ver, mas como vimos isso ajuda a reduzir significamente a quantidade de código de lógica de UI de nossas viewcontrollers.
+Desta forma toda vez que a property dateAdded do nosso modelo, for modificada a property dateAdded do modelo será populada com o valor correspondente. E nós criamos uma conexão "invisível" entre elas. Essa é a razão pela qual chamamos reativa: duas properties estão agora relacionadas, então as mudanças que ocorrerem em uma, irá afetar reativamente a outra.    
```

Comment:
"do nosso modelo for modificada, a property" 
"então as mudanças que ocorrerem em uma irão afetar reativamente a outra."